### DISCUSSION TITLE
How to setup Ruby language as in Atom?

### DISCUSSION THREAD

**Discussion by amigobg** [pulsar-edit/.github#234] [Pulsar Support]
Could you guide me on configuring Pulsar to match the Ruby syntax coloring used in Atom? The attached image shows the current syntax display in Pulsar on the left, compared to Atom's syntax on the right. Both are utilizing the 'language-ruby' and 'language-ruby-on-rails' packages, and neither has undergone additional color customization.

![difference](https://github.com/pulsar-edit/.github/assets/16067820/e4506216-4e64-4a3b-bebb-19c21e5c71ce)

**Comment by Daeraxa** [comment]
In Settings > Core do you have the following enabled or disabled?
![image](https://github.com/pulsar-edit/.github/assets/58074586/d8cd5c70-2ac6-414d-81e6-c89ed5a38797)

If it is enabled then this is the new, replacement system where there are (or were) some running fixes going on based on feedback. @savetheclocktower has been looking at this based on previous feedback, not 100% sure where things are with this. What version are you running?

**Reply by amigobg** [reply]
Was disabled, now syntax seems to be the same as in Atom.

**Reply by Daeraxa** [reply]
That's odd... If it is disabled then I would have expected it to be the same as Atom, if enabled then we do have differences because of the way the languages are parsed - sometimes different for the better.

**Reply by Daeraxa** [reply]
And what version are you running?

**Reply by amigobg** [reply]
1.111.0 Actually now when is enabled is not the same as Atom but its better then the screenshot where is disabled.

**Reply by Daeraxa** [reply]
The fact it looks different when disabled is more of a worry as we shouldn't have made any changes to how the previous tree-sitter grammar displays stuff, I thought anyway...

Something is definitely different between them when you compare both legacy tree-sitter grammars.

Atom 1.60.0 language-ruby tree-sitter
```html
<div class="line" data-screen-row="0">
  <span class="">
    <span class="syntax--source syntax--ruby">
      <span class="syntax--entity syntax--name syntax--function">
        store_accessor
      </span>
      <span class="syntax--constant syntax--other syntax--symbol">:params</span>
      <span class="syntax--punctuation syntax--separator syntax--object">
        ,
      </span>{" "}
      <span class="syntax--constant syntax--other syntax--symbol">
        :from_date
      </span>
      <span class="syntax--punctuation syntax--separator syntax--object">
        ,
      </span>
      <span class="syntax--constant syntax--other syntax--symbol">
        :to_date
      </span>
    </span>
  </span>
</div>;

```

Pulsar 1.111.0 language-ruby **legacy** tree-sitter
```html
<div class="line" data-screen-row="0">
  <span class="">
    <span class="syntax--source syntax--ruby">
      store_accessor :params
      <span class="syntax--punctuation syntax--separator syntax--object">
        ,
      </span>{" "}
      :from_date
      <span class="syntax--punctuation syntax--separator syntax--object">
        ,
      </span>{" "}
      :to_date
    </span>
  </span>
</div>;

```

Pulsar 1.111.0 language-ruby **modern** tree-sitter (not technically related, just for sake of completeness)
```html
<div class="line" data-screen-row="0">
  <span class="">
    <span class="syntax--source syntax--ruby">
      <span class="syntax--support syntax--other syntax--function syntax--ruby">
        store_accessor
      </span>
      <span class="syntax--constant syntax--other syntax--symbol syntax--ruby">
        <span class="syntax--punctuation syntax--definition syntax--symbol syntax--ruby">
          :
        </span>
        params
      </span>
      <span class="syntax--punctuation syntax--separator syntax--comma syntax--ruby">
        ,
      </span>
      <span class="syntax--constant syntax--other syntax--symbol syntax--ruby">
        <span class="syntax--punctuation syntax--definition syntax--symbol syntax--ruby">
          :
        </span>
        from_date
      </span>
      <span class="syntax--punctuation syntax--separator syntax--comma syntax--ruby">
        ,
      </span>
      <span class="syntax--constant syntax--other syntax--symbol syntax--ruby">
        <span class="syntax--punctuation syntax--definition syntax--symbol syntax--ruby">
          :
        </span>
        to_date
      </span>
    </span>
  </span>
</div>;

```

**Reply by savetheclocktower** [reply]
> 1.111.0 Actually now when is enabled is not the same as Atom but its better then the screenshot where is disabled.

Did you attempt to migrate your settings from your Atom installation, or did you start fresh with default settings? My theory is that you had disabled Tree-sitter grammars at some point back when you were using Atom.

The new Tree-sitter system will definitely look _slightly_ different. We expect people to think “this will take a little getting used to,” rather than “this makes no sense and is not helpful.” If there are any differences that feel outright wrong, please let us know.

If you want the ability to use the legacy TextMate-style Ruby grammar, follow the instructions in the second paragraph of [this section](https://pulsar-edit.dev/blog/20230925-savetheclocktower-modern-tree-sitter-part-1.html#can-i-use-this-new-implementation-now) to be able to switch between the three different Ruby grammars we currently have. After you do this, you'll see three Ruby grammars in the grammar selector: one marked “Tree-sitter,” one marked “Legacy Tree-sitter,” and one without any badge at all. The third one might be what you're expecting.

If it is, and if you vastly prefer the old grammar's syntax highlighting, here's what we'd prefer you do:

* Ask us about specific changes you'd like to see — e.g., “I expect the colon in a symbol to be a different color from the rest of the symbol.” We might be able to make the change in the grammar.
* If we think the suggested change doesn't make sense, we can usually still tell you how to override your syntax theme to achieve what you want without having to change the grammar you're using.
* If that doesn't work and you prefer the old TextMate-style grammar, we can tell you how to disable the newer Tree-sitter grammar _for Ruby specifically_ without having to turn off Tree-sitter altogether.

**Reply by savetheclocktower** [reply]
> Did you attempt to migrate your settings from your Atom installation, or did you start fresh with default settings? My theory is that you had disabled Tree-sitter grammars at some point back when you were using Atom.

Never mind — @Daeraxa helped me realize what happened here. We upgraded all the legacy Tree-sitter parsers from the versions that were present in Atom, and that introduced some breaking changes.

Luckily, the fix is to do what you've already done: enable the experimental new Tree-sitter system, which will very soon be the _only_ Tree-sitter system.
